### PR TITLE
 Build and Secure an API in Python with FastAPI 

### PR DESCRIPTION
**Pull request by jpf**
This blog post has the following features:

- [X] A GitHub Repository with a polished README
- [X] A GitHub Repository under the github.com/oktadeveloper account
- [X] A title that's approved by Dev Advocacy
- [X] A URL that's approved by Dev Advocacy
- [X] Rendered locally and confirmed that no Markdown typos exist

### CODE REVIEW

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
+---
+layout: blog_post
+title: "Build and Secure an API in Python with FastAPI"
+author: karl-hughes
+by: contractor
+communities: [python]
+description: "How to quickly build a REST API with FastAPI and secure it with Okta"
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:7]:
```suggestion
description: "How to quickly build a Python REST API with FastAPI and secure it with Okta."
```

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+tags: [python, fastapi, authentication]
+tweets:
+- "Learn how to quickly build a REST API in Python with FastAPI"
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:10]:
```suggestion
- "Learn how to quickly build a REST API in Python with FastAPI!"
```

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+- "Tired of the overhead it takes to build a REST API? Use FastAPI!"
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:11]:
```suggestion
- "Tired of the overhead it takes to build a REST API with Python? Use FastAPI!"
```

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+- "Quickly build out a database backed REST API with documentation using FastAPI"
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:12]:
```suggestion
- "Quickly build out a database-backed Python REST API with documentation using FastAPI."
```

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+First released [in late 2018](https://github.com/tiangolo/fastapi/releases?after=0.2.0), FastAPI differentiates itself from other Python frameworks by offering a modern, fast, and succinct developer experience for building reliable [REST APIs](/blog/2019/09/04/securing-rest-apis). While one of the newer open-source Python frameworks available, FastAPI has quickly gained a following with over [22,000 stars on GitHub](https://github.com/tiangolo/fastapi) and an active [community of maintainers](https://fastapi.tiangolo.com/fastapi-people/) working on the project.
+
+In this post, you'll learn more about FastAPI, and why it might be the right choice for your next project. You'll see how to set up a new FastAPI project and use [Okta](https://www.okta.com/) to secure the API.
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:20]: In previous times, Lindsay used to recommend we have no links in the opening paragraphs so people don't leave the post before they read it. It's probably too late to do it in this post, but I'd try to keep it in mind for future posts.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+Finally, while FastAPI comes with many of the features you would expect in a REST API framework (like [data validation](https://fastapi.tiangolo.com/features/#validation) and [authentication](https://fastapi.tiangolo.com/features/#security-and-authentication)), it lets you choose your ORM and database of choice. This means that FastAPI can work with your existing data models if you're migrating from an existing Python application.
+
+While FastAPI is an excellent option for building REST APIs in Python, it's not perfect for every situation. If the system you're building relies on Python 3.5 or any earlier Python framework, you won't be able to use FastAPI. FastAPI also isn't a great choice if you're building a server-side web application that needs to display HTML, as it doesn't include all the view helper functions that full-stack frameworks like Django do.
+
+Now that you're familiar with FastAPI at a high level, you're ready to start building your first application. In the rest of this tutorial, you'll see how to get started and secure your endpoints using Okta as your OAuth authorization server.
+
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:37]: I like to do a prerequisites and table of contents right at the beginning. Here's a sample:

```
**Prerequisites**:

- [Node 14](https://nodejs.org/)
- [Okta CLI 0.7.1+](https://github.com/okta/okta-cli)

**Table of Contents**{: .hide }
* Table of Contents
{:toc}
```

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+To start your application, run the Uvicorn server:
+
+```bash
+uvicorn main:app --reload
+```
+
+Then, open [http://127.0.0.1:8000](http://127.0.0.1:8000) in your browser to see your "Hello, world" endpoint in action:
+
+![A "Hello, World" endpoint in FastAPI](https://i.imgur.com/XHwSIuy.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:87]: This should be an image embedded in the post, so we can optimize it.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+
+When using FastAPI, there are two ways you can access your API. You can use an API explorer like [Postman](https://www.postman.com/), or you can use FastAPI's interactive documentation in your browser. This tutorial will focus on the latter.
+
+To start your application, run the Uvicorn server:
+
+```bash
+uvicorn main:app --reload
+```
+
+Then, open [http://127.0.0.1:8000](http://127.0.0.1:8000) in your browser to see your "Hello, world" endpoint in action:
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:85]:
```suggestion
Then, open `http://127.0.0.1:8000` in your browser to see your "Hello, world" endpoint in action:
```

We don't typically link to localhost - we use backticks around them instead. It comes from before times when we'd validate all links. It wouldn't be a bad idea to start validating them again, at least weekly. I've heard https://www.drlinkcheck.com/ works well.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+![A "Hello, World" endpoint in FastAPI](https://i.imgur.com/XHwSIuy.png)
+
+FastAPI's interactive documentation is available at [http://127.0.0.1:8000/docs](http://127.0.0.1:8000/docs). Once you're there, you'll see details about each endpoint automatically. FastAPI uses type hints and context in your application to produce these docs on its own.
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:89]: Backticks around localhost links.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+![FastAPI interactive documentation](https://i.imgur.com/Va63heZ.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:91]: Embed images so they can be optimized and cached appropriately.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+
+Now that you've seen how easy it is to get started, you're ready to build a more useful application. In the remainder of this tutorial, you'll see how to create protected endpoints in FastAPI that use Okta as an authorization server.
+
+## Securing FastAPI with Okta
+
+To demonstrate some of the more advanced features of the FastAPI framework, I'll show you how to build a protected endpoint that uses [the client credentials flow](https://developer.okta.com/docs/guides/implement-client-creds/overview/) to authorize access. By the end of this tutorial, you will be able to enter your Okta client ID and secret into FastAPI's interactive docs to get an access token. You'll see how to validate this JWT access token remotely using Okta's [introspect](https://developer.okta.com/docs/reference/api/oidc/#introspect) endpoint and locally using the [Okta JWT Python library](https://github.com/adithyasampatoor/okta_jwt).
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:97]: Links to d.o.c should be relative so they work in preview. Also, is there no official Okta Python library?

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+
+If you'd like to run the final application, [the code is available on GitHub](https://github.com/oktadeveloper/okta-fastapi), or you can follow along for step-by-step instructions.
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:99]: We try to prefix our blog post repos with "okta" and suffix them with "example". I also like putting the keywords in the repo name, so maybe this should be "okta-python-fastapi-example"?

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+
+### Setting Up a New Application in Okta
+
+Before you create any endpoints in your FastAPI application, you'll need to create a new application in Okta and get your Authorization Server's issuer URL and audience. As you go through these steps, add the Okta environment variables to a new file in your application called `.env`. You'll see how to use these variables later in the tutorial.
+
+To create a new server application, log in to your Okta account and go to **Applications** and click the **Add Application** button in the top left. Select **Service, Machine-to-Machine**, then click **Next**.
+
+![Creating a new service application in Okta](https://i.imgur.com/8EUhPMx.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:107]: Embed

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+
+Enter a name for your application and click **Next** again.
+
+![Entering a name for your application in Okta](https://i.imgur.com/IS1Z4jn.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:111]: Embed

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+
+Copy the **Client ID** and **Client Secret** from this page and add them to your FastAPI application's `.env` file as `OKTA_CLIENT_ID` and `OKTA_CLIENT_SECRET` respectively.
+
+Your `.env` file should look like the example below, with your `OKTA_CLIENT_ID` and `OKTA_CLIENT_SECRET` values filled out:
+
+```
+OKTA_CLIENT_ID=
+OKTA_CLIENT_SECRET=
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:119]: We typically do placeholders like `{yourClientID}`, but it's probably fine to leave like this.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+``` 
+
+![Retrieving your client credentials in Okta](https://i.imgur.com/dSVWopB.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:122]: Embed image.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+
+Next, go to **API > Authorization Servers**. Copy the **Issuer URI** and **Audience**, and add them as the `OKTA_ISSUER` and `OKTA_AUDIENCE` environment variables in your `.env` file.
+
+Your `.env` file should now look something like this:
+
+```
+OKTA_CLIENT_ID=
+OKTA_CLIENT_SECRET=
+OKTA_ISSUER="https://SOMETHING_HERE.oktapreview.com/oauth2/default"
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:131]: New dev orgs are created on okta.com, so I'd change this to reflect that.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+OKTA_AUDIENCE="api://default"
+```
+
+![Retrieving your Okta Issuer URI](https://i.imgur.com/7VPIjKV.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:135]: Embed.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+
+Now that you have your environment variables, you need to create a custom scope for your authorization server. Your FastAPI application will request a token with this scope. Click the pencil icon to edit the authorization server.
+
+![Editing your Okta authorization server](https://i.imgur.com/8G5Y6dk.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:139]: Embed.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+
+Click the **Scopes** tab and then the **Add Scopes** button.
+
+![Adding a scope to your Okta authorization server, step 1](https://i.imgur.com/j04IktM.png)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:143]: Embed. You get the point. ;)

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+        )
+        return bool(res)
+    except Exception:
+        raise HTTPException(status_code=403)
+...
+```
+
+When you call the `/items` endpoint, the API will decode the JWT and validate it locally. The decoded JWTs are cached, so subsequent requests will be faster than the first one.
+
+## Conclusion
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:349]: I'd change this to "Learn More About Python and REST APIs" so it's a better SEO header.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+
+If you want to learn more about FastAPI, I suggest the following resources:
+- [Fast API from the ground up (video)](https://www.youtube.com/watch?v=3DLwPcrE5mA)
+- [Basic OAuth in FastAPI](https://fastapi.tiangolo.com/tutorial/security/simple-oauth2)
+- [Developing a CRUD API with FastAPI](https://testdriven.io/blog/fastapi-crud/)
+
+To learn more about building an app that can communicate with FastAPI, I suggest using one of these tutorials:
+- [Data Visualization in Angular Using D3.js](https://blog.logrocket.com/data-visualization-angular-d3/)
+- [Building Micro Frontends](https://www.telerik.com/blogs/building-micro-frontends)
+- [Building a Response Timer to Benchmark API Performance](https://dev.to/karllhughes/building-a-response-timer-to-benchmark-api-performance-3k6k)
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:362]: Missing a call to action for following `@oktadev` and links to other Python / REST posts.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+
+Finally, while FastAPI comes with many of the features you would expect in a REST API framework (like [data validation](https://fastapi.tiangolo.com/features/#validation) and [authentication](https://fastapi.tiangolo.com/features/#security-and-authentication)), it lets you choose your ORM and database of choice. This means that FastAPI can work with your existing data models if you're migrating from an existing Python application.
+
+While FastAPI is an excellent option for building REST APIs in Python, it's not perfect for every situation. If the system you're building relies on Python 3.5 or any earlier Python framework, you won't be able to use FastAPI. FastAPI also isn't a great choice if you're building a server-side web application that needs to display HTML, as it doesn't include all the view helper functions that full-stack frameworks like Django do.
+
+Now that you're familiar with FastAPI at a high level, you're ready to start building your first application. In the rest of this tutorial, you'll see how to get started and secure your endpoints using Okta as your OAuth authorization server.
+
+## Building a New FastAPI Project
```
Comment (mraible) [inline _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:38]: I like that FastAPI is in the first couple headers for SEO. Is FastAPI the only keyword or should "Python" be added?

**Review by mraible — CHANGES_REQUESTED** [review]

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md (hunk 5 above)
Comment (jpf) [reply _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md:20]: Good to know. Thanks.

File: _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md (hunk 23 above)
Comment (jpf) [reply _source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md]: Good catch, FastAPI is fine here as "Python" is a highly competitive keyword.

**Comment by jpf** [comment]
@mraible how does this look now? I addressed all of your feedback aside from the comments still visible above.

**Review by mraible — APPROVED** [review]

**Pull request merged**